consensus/phases/thirdphase.go

```diff
@@ -129,6 +131,36 @@ func (tp *ThirdPhaseImpl) Execute(ctx context.Context, pulse *core.Pulse, state
 	}, nil
 }
 
+func (tp *ThirdPhaseImpl) handleJoinClaims(ref core.RecordRef, state *SecondPhaseState, entropy core.Entropy) {
+	handler := claimhandler.NewJoinHandler(len(tp.NodeKeeper.GetActiveNodes()))
+	claims := state.UnsyncList.GetClaims(ref)
+
+	if len(claims) == 0 {
+		return
+	}
+
+	joinClaims := make([]*packets.NodeJoinClaim, 0)
+	resClaims := make([]packets.ReferendumClaim, 0)
+	logger := inslogger.FromContext(context.Background())
+	logger.Debugf("[ handleJoinClaims ] old claims count: %d", len(claims))
+	for _, claim := range claims {
+		c, ok := claim.(*packets.NodeJoinClaim)
+		if !ok {
+			resClaims = append(resClaims, c)
+			continue
+		}
+		joinClaims = append(joinClaims, c)
+	}
+	updatedJoinClaims := handler.HandleClaims(joinClaims, entropy)
+	if len(updatedJoinClaims) > 0 {
+		for _, claim := range updatedJoinClaims {
+			resClaims = append(resClaims, claim)
+		}
+		logger.Debugf("[ handleJoinClaims ] new claims count: %d", len(resClaims))
+		state.UnsyncList.UpdateClaims(ref, resClaims)
```

Comment:
I think we can have a problem here because we do not remove claims, only update

Reply:
yep cuz we have to insert a claims in this func if and only if a join claims count was changed. look at the resultClaims. its a claim slice w/o removed claims

Reply:
ok, i got it